kusto/ingest/status.go

```diff
@@ -0,0 +1,301 @@
+	safeSetString(data, "ErrorCode", &r.ErrorCode)
+	safeSetString(data, "Details", &r.Details)
+
+	if data["IngestionSourceId"] != nil {
+		if uid, err := getGoogleUUIDFromInterface(data["IngestionSourceId"]); err == nil {
+			r.IngestionSourceID = uid
+		}
+	}
+
+	if data["OperationId"] != nil {
+		if uid, err := getGoogleUUIDFromInterface(data["OperationId"]); err == nil {
+			r.OperationID = uid
+		}
+	}
+
+	if data["ActivityId"] != nil {
+		if uid, err := getGoogleUUIDFromInterface(data["ActivityId"]); err == nil {
+			r.ActivityID = uid
+		}
+	}
+
+	if data["UpdatedOn"] != nil {
+		if t, err := getTimeFromInterface(data["UpdatedOn"]); err == nil {
+			r.UpdatedOn = t
+		}
+	}
+
+	if data["OriginatesFromUpdatePolicy"] != nil {
+		if b, ok := data["OriginatesFromUpdatePolicy"].(bool); ok {
+			r.OriginatesFromUpdatePolicy = b
+		}
+	}
+}
+
+// ToMap converts an ingestion status record to a key value map.
+func (r *statusRecord) ToMap() map[string]interface{} {
+	data := make(map[string]interface{})
+
+	// Since we only create the initial record, It's not our responsibility to write the following fields:
+	//   OperationID, AcitivityID, ErrorCode, FailureStatus, Details, OriginatesFromUpdatePolicy
+	// Those will be read from the server if they have data in them
+	data["Status"] = r.Status
+	data["IngestionSourceId"] = r.IngestionSourceID
+	data["IngestionSourcePath"] = r.IngestionSourcePath
+	data["Database"] = r.Database
+	data["Table"] = r.Table
+	data["UpdatedOn"] = r.UpdatedOn.Format(time.RFC3339Nano)
+
+	return data
+}
+
+// String implements fmt.Stringer.
+func (r *statusRecord) String() string {
+	return pretty.Sprint(r)
+}
+
+// ToError converts an ingestion status to an error if failed or partially succeeded, or nil if succeeded.
+func (r statusRecord) Error() string {
+	switch r.Status {
+	case Succeeded:
+		return fmt.Sprintf("Ingestion succeeded\n" + r.String())
+
+	case Queued:
+		return fmt.Sprintf("Ingestion Queued\n" + r.String())
+
+	case PartiallySucceeded:
+		return fmt.Sprintf("Ingestion succeeded partially\n" + r.String())
+
+	default:
+		return fmt.Sprintf("Ingestion Failed\n" + r.String())
+	}
+}
+
+func getTimeFromInterface(x interface{}) (time.Time, error) {
+	switch x.(type) {
+	case string:
+		return time.Parse(time.RFC3339Nano, x.(string))
+
+	case time.Time:
+		return x.(time.Time), nil
+
+	default:
+		return time.Now(), fmt.Errorf("getTimeFromInterface: Unexpected format %T", x)
+	}
+}
+
+func getGoogleUUIDFromInterface(x interface{}) (uuid.UUID, error) {
+	switch x.(type) {
+	case string:
+		return uuid.Parse(x.(string))
+
+	case uuid.UUID:
+		return x.(uuid.UUID), nil
+
+	case storageuid.UUID:
+		uid, err := uuid.ParseBytes(x.(storageuid.UUID).Bytes())
+		if err != nil {
+			return uuid.Nil, err
+		}
+
+		return uid, err
+
+	default:
+		return uuid.Nil, fmt.Errorf("getGoogleUUIDFromInterface: Unexpected format %T", x)
+	}
+}
+
+func safeSetString(data map[string]interface{}, key string, target *string) {

```

Comment:
I think we should change how we are doing this to:

```go
func extractKey(m map[string]interface{}, key string) string {
  if v := data[key]; v != nil {
    if s, ok := v.(string); ok {
      return s
    }
  }
  return ""
}
```

Then above we do:

```go
r.IngestionSourcePath = extractKey(data, "IngestionSourcePath")
```
This keeps the operations as string copies, no pointer conversions and prevents any possibility of escape analysis that could have us putting things on the heap.